sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/UnwrapCastInBinaryComparison.scala

```diff
@@ -329,6 +334,40 @@ object UnwrapCastInBinaryComparison extends Rule[LogicalPlan] {
     }
   }
 
+  private def unwrapTimeStampToDate(
+      exp: BinaryComparison,
+      fromExp: Expression,
+      ts: Literal,
+      tz: Option[String],
+      evalMode: EvalMode.Value): Expression = {
+    val floorDate = Cast(ts, fromExp.dataType, tz, evalMode)
+    val equalToFloorDateExpr =
+      EqualTo(ts, Cast(floorDate, ts.dataType, tz, evalMode))
+    val equalToFloorDate = equalToFloorDateExpr.eval(EmptyRow).asInstanceOf[Boolean]
+    val ceilDate =
+      if (equalToFloorDate) {
+        floorDate
+      } else {
+        DateAdd(floorDate, Literal(1, IntegerType))
+      }
+
+    exp match {
+      case _: GreaterThan =>
+        GreaterThan(fromExp, floorDate)
+      case _: GreaterThanOrEqual =>
+        GreaterThanOrEqual(fromExp, ceilDate)
+      case _: EqualTo =>
+        And(EqualTo(fromExp, floorDate), EqualTo(fromExp, ceilDate))
+      case EqualNullSafe(_, _) =>
```

Comment:
It seems the left side should not be nullable.
```suggestion
      case EqualNullSafe(left, _) if !left.nullable =>
```

Reply:
How about?
```scala
        if (isStartOfDay) EqualNullSafe(fromExp, floorDate) else FalseLiteral
```

Reply:
Updated, thanks